# problem statement 

### 1.IMPORT REQUIRED LIBRARYS



In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os 
from sklearn.model_selection import train_test_split
import pickle

In [2]:
 pd.set_option('display.float_format', '{:.2f}'.format)

###  2.READING THE DATA 

In [3]:
def get_data():

    #Noida  city csv file:
    #jaipur=pd.read_csv('jaipur.csv')
    
    #Noida  city csv file:
    #bhopal=pd.read_csv('bhopal.csv')

    
    #Noida  city csv file:
    #vadodara=pd.read_csv('vadodara.csv')

    
    #Noida  city csv file:
    #surat=pd.read_csv('surat.csv')

    
    #Noida  city csv file:
    #kanpur=pd.read_csv('kanpur.csv')

    
    #Noida  city csv file:
    #Guwahati=pd.read_csv('Guwahati.csv')

     #Noida  city csv file:
    #Faridabad=pd.read_csv('Faridabad.csv')

    


    

    #Noida  city csv file:
    #Lucknow=pd.read_csv('Lucknow.csv')

    #indore  city csv file:
    #indore=pd.read_csv('indore.csv')

    #Noida  city csv file:
    Noida=pd.read_csv('Noida.csv')

    #Ghaziyabad  city csv file:
    #Ghaziyabad=pd.read_csv('Ghaziyabad.csv')

    # Ahmadabad city csv file:
    Ahmadabad=pd.read_csv('Ahmadabad.csv')
    

    # kolkata city csv file:
    kolkata=pd.read_csv('Kolkata.csv')

    
    # Hydrabadcity csv file:
    hyderabad=pd.read_csv('Hydrabad.csv')
    
    
    # gurgaoncity csv file:
    gurgaon=pd.read_csv('Gurgaon.csv')

    # Delhi city csv file:
    Delhi=pd.read_csv('Delhi.csv')
    
    # Chennai city csv file:
    Chennai=pd.read_csv('Chennai.csv')

    # Banglore city csv file:
    Banglore=pd.read_csv('Banglore.csv')
    
    # pune city csv file:
    pune=pd.read_csv('Pune.csv')
    
    # mumbai city csv file:
    mumbai=pd.read_csv('Mumbai.csv')

   ### merge alll above city into one dataframe:
    return  pd.concat([mumbai,pune,Banglore,Chennai,Delhi,gurgaon,hyderabad,kolkata,Ahmadabad,Noida],ignore_index=True)
    #return  pd.concat([mumbai,pune,Banglore,Chennai,Delhi,gurgaon,hyderabad,kolkata,Ahmadabad,Ghaziyabad,Noida,indore,Lucknow,jaipur,bhopal,vadodara,surat,kanpur,Guwahati,Faridabad],ignore_index=True
    
    

In [4]:
data=get_data()

In [5]:
data

broker_consultancy   Rent  Area  appartment_type  \
0      Kasturi Developers\nBUILDER\n-  20000  1132                2   
1      Kasturi Developers\nBUILDER\n-  20000  1200                2   
2      Kasturi Developers\nBUILDER\n-  30000  1700                3   
3              Seller\nVERIFIED OWNER  28500   750                2   
4              Seller\nVERIFIED OWNER  50000  1400                3   
...                               ...    ...   ...              ...   
74891          Seller\nVERIFIED OWNER  60000  1880                3   
74892          Seller\nVERIFIED OWNER  15000   500                2   
74893          Seller\nVERIFIED OWNER   9000   150                1   
74894          Seller\nVERIFIED OWNER   9200   300                1   
74895          Seller\nVERIFIED OWNER  15000   225                1   

              Deposit    Bathrooms            Facing                 Address  \
0          No Deposit  2 bathrooms  NorthEast facing            Ulwe, Mumbai   
1          No Deposit  2 bathrooms  NorthEast facing            Ulwe, Mumbai   
2          No Deposit  3 bathrooms  NorthEast facing            Ulwe, Mumbai   
3          No Deposit  2 bathrooms               NaN  Kandivali West, Mumbai   
4          No Deposit  3 bathrooms       East facing         Chembur, Mumbai   
...               ...          ...               ...                     ...   
74891   1.2 L deposit  3 bathrooms               NaN        Sector 79, Noida   
74892  15,000 deposit  2 bathrooms               NaN       Sector 128, Noida   
74893      No Deposit  1 bathrooms  NorthEast facing        Sector 52, Noida   
74894   4,000 deposit  1 bathrooms               NaN       Sector 135, Noida   
74895  10,000 deposit  1 bathrooms               NaN        Sector 44, Noida   

          Apartment Name          Status  Rent/sqfit localityclass  
0              EV Castle     Unfurnished       17.67        class2  
1                    NaN     Unfurnished       16.67        class2  
2                    NaN  Semi-Furnished       17.65        class2  
3         Atul Blue Arch     Unfurnished       38.00        class2  
4           Godrej Prime  Semi-Furnished       35.71        class2  
...                  ...             ...         ...           ...  
74891  Mahagun Mirabella  Semi-Furnished       31.91        class1  
74892                NaN  Semi-Furnished       30.00        class2  
74893                NaN       Furnished       60.00        class1  
74894                NaN       Furnished       30.67        class1  
74895                NaN       Furnished       66.67        class1  

[74896 rows x 12 columns]

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74896 entries, 0 to 74895
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   broker_consultancy  74895 non-null  object 
 1   Rent                74896 non-null  int64  
 2   Area                74896 non-null  int64  
 3   appartment_type     74896 non-null  int64  
 4   Deposit             74896 non-null  object 
 5   Bathrooms           74552 non-null  object 
 6   Facing              26376 non-null  object 
 7   Address             74896 non-null  object 
 8   Apartment Name      30357 non-null  object 
 9   Status              74896 non-null  object 
 10  Rent/sqfit          74896 non-null  float64
 11  localityclass       74896 non-null  object 
dtypes: float64(1), int64(3), object(8)
memory usage: 6.9+ MB


-   The data set contains 82,016 rows and 10 features
-   column "Bathrooms ,Facing,Appartment_Name," have lots of  missing values
-   The data type of some features are not appropriate

### 3.PREMLINARY DATA ANALYSIS

#### 3.1Check for datatype


In [7]:
data.sample(150)

broker_consultancy   Rent  Area  appartment_type  \
27233                 Seller\nVERIFIED OWNER  13000  2000                1   
4236      Shabbir Precise Property\nAGENT\n-  42000  1100                2   
20920                 Seller\nVERIFIED OWNER  35000  1500                3   
33941                 Seller\nVERIFIED OWNER  35000  1200                3   
31417                 Seller\nVERIFIED OWNER   9500  1200                2   
...                                      ...    ...   ...              ...   
70961                 Seller\nVERIFIED OWNER  25000  1400                2   
41297                 Seller\nVERIFIED OWNER  48000  1020                3   
15413  Prince Property Navi Mumbai\nAGENT\n-  12000   635                1   
29658                 Seller\nVERIFIED OWNER  20000   550                1   
8530            Sunrise Properties\nAGENT\n-  72000  1200                2   

              Deposit    Bathrooms            Facing  \
27233      No Deposit  3 bathrooms               NaN   
4236       No Deposit  2 bathrooms               NaN   
20920      No Deposit  2 bathrooms      North facing   
33941      No Deposit  2 bathrooms               NaN   
31417      No Deposit  2 bathrooms  NorthEast facing   
...               ...          ...               ...   
70961      No Deposit  2 bathrooms       East facing   
41297      No Deposit  3 bathrooms      South facing   
15413  35,000 deposit  2 bathrooms               NaN   
29658      No Deposit  1 bathrooms       West facing   
8530       No Deposit  2 bathrooms       East facing   

                            Address  \
27233          Gottigere, Bangalore   
4236        Jogeshwari West, Mumbai   
20920              Sangamvadi, Pune   
33941         Swasthya Vihar, Delhi   
31417  Kithaganur Colony, Bangalore   
...                             ...   
70961             Shilaj, Ahmedabad   
41297             Model Town, Delhi   
15413    Sector 21 Ghansoli, Mumbai   
29658         Whitefield, Bangalore   
8530             Malad East, Mumbai   

                                  Apartment Name          Status  Rent/sqfit  \
27233                                        NaN  Semi-Furnished        6.50   
4236   Aftab Shagufa Municipal Employees CHS Ltd     Unfurnished       38.18   
20920                                        NaN       Furnished       23.33   
33941                                        NaN       Furnished       29.17   
31417                                        NaN  Semi-Furnished        7.92   
...                                          ...             ...         ...   
70961                            Shafalya Renown       Furnished       17.86   
41297                                        NaN       Furnished       47.06   
15413       Prince Property Ghansoli Navi Mumbai     Unfurnished       18.90   
29658                                        NaN       Furnished       36.36   
8530                            DB Shagun Towers       Furnished       60.00   

      localityclass  
27233        class3  
4236         class2  
20920        class2  
33941        class2  
31417        class3  
...             ...  
70961        class2  
41297        class1  
15413        class2  
29658        class1  
8530         class1  

[150 rows x 12 columns]

In [8]:
data.dtypes

broker_consultancy     object
Rent                    int64
Area                    int64
appartment_type         int64
Deposit                object
Bathrooms              object
Facing                 object
Address                object
Apartment Name         object
Status                 object
Rent/sqfit            float64
localityclass          object
dtype: object

#### 3.2Check For Duplicates

In [9]:
data.duplicated().sum()

62121

In [10]:
duplicated_rows=(
    data.loc[data.duplicated(keep=False)].sort_values(["Rent","Area","appartment_type","Bathrooms","Facing","Address","Apartment Name","Status"])
)

In [11]:
duplicated_rows.sample(150)

broker_consultancy   Rent  Area  appartment_type  \
17850                 Vision Realty\nAGENT\n-  30000  1570                3   
33461                  Seller\nVERIFIED OWNER  12500  1500                3   
44606              World Wide Infra\nAGENT\n-  90000  2800                4   
11074                 Prema Housing\nAGENT\n-  45000  1020                2   
53966                  Seller\nVERIFIED OWNER  60000  2300                4   
...                                       ...    ...   ...              ...   
49849                  Seller\nVERIFIED OWNER  13000  1000                1   
63576                  Seller\nVERIFIED OWNER  23000  1154                3   
46045        Mahadev Renting Expert\nAGENT\n0  32500  1100                3   
31871                     HomeFirst\nAGENT\n-  80000  2250                3   
33492  Nimmadhi Property Management\nAGENT\n-  31000  2713                3   

           Deposit    Bathrooms            Facing  \
17850   No Deposit  3 bathrooms               NaN   
33461   No Deposit  3 bathrooms               NaN   
44606   No Deposit  4 bathrooms               NaN   
11074  2 L deposit  2 bathrooms               NaN   
53966   No Deposit  4 bathrooms       East facing   
...            ...          ...               ...   
49849   No Deposit  1 bathrooms               NaN   
63576   No Deposit  2 bathrooms  SouthWest facing   
46045   No Deposit  2 bathrooms               NaN   
31871   No Deposit  3 bathrooms  SouthWest facing   
33492   No Deposit  3 bathrooms               NaN   

                             Address             Apartment Name  \
17850  NIBM Annex Mohammadwadi, Pune    Kumar Palmspring Towers   
33461             Urapakkam, Chennai                        NaN   
44606             Vasant Kunj, Delhi  DDA DDA Sector B Pocket 9   
11074          Borivali East, Mumbai        Gurukrupa Raj Hills   
53966             Sector 24, Gurgaon        DLF Belvedere Tower   
...                              ...                        ...   
49849              Sector 4, Gurgaon    HUDA RWA Swaran Jayanti   
63576              Rajarhat, Kolkata           Aspira Aashiyana   
46045               Janakpuri, Delhi         DDA Flat Janakpuri   
31871          Nungambakkam, Chennai                        NaN   
33492                 Porur, Chennai                        NaN   

               Status  Rent/sqfit localityclass  
17850  Semi-Furnished       19.11        class2  
33461  Semi-Furnished        8.33        class3  
44606  Semi-Furnished       32.14        class2  
11074     Unfurnished       44.12        class2  
53966  Semi-Furnished       26.09        class1  
...               ...         ...           ...  
49849  Semi-Furnished       13.00        class3  
63576       Furnished       19.93        class2  
46045  Semi-Furnished       29.55        class2  
31871       Furnished       35.56        class1  
33492  Semi-Furnished       11.43        class3  

[150 rows x 12 columns]

In [12]:
duplicated_rows.shape

(64547, 12)

In [13]:
82016-68994

13022

#### 3.3 Observations

- The type of `Rent`, `Deposit`, `bathroom`should be changed to numeric type.
- There're 6700 duplicates. These should be removed.

### 4.DETAILED ANALYSIS

#### broker_consultancy


In [14]:
data.broker_consultancy

0        Kasturi Developers\nBUILDER\n-
1        Kasturi Developers\nBUILDER\n-
2        Kasturi Developers\nBUILDER\n-
3                Seller\nVERIFIED OWNER
4                Seller\nVERIFIED OWNER
                      ...              
74891            Seller\nVERIFIED OWNER
74892            Seller\nVERIFIED OWNER
74893            Seller\nVERIFIED OWNER
74894            Seller\nVERIFIED OWNER
74895            Seller\nVERIFIED OWNER
Name: broker_consultancy, Length: 74896, dtype: object

In [15]:
data.broker_consultancy.dtypes

dtype('O')

In [16]:
data.broker_consultancy.isna().sum()

1

In [17]:
data.broker_consultancy.unique()

array(['Kasturi Developers\nBUILDER\n-', 'Seller\nVERIFIED OWNER',
       'CHERRISH ESTATE CONSULTANTS\nAGENT\n-', ...,
       'City Estate Management\nAGENT\n-', 'Daata Real Estate\nAGENT\n0',
       'City Estate Management\nAGENT\n0'], dtype=object)

-   The data type of broker_consultancy is valid
-   broker_consultancy contain one null value
-   There is no used of this feature in model_trainng
-   we are going drop this in data_cleaning operation




#### Rent

In [18]:
data.Rent

0        20000
1        20000
2        30000
3        28500
4        50000
         ...  
74891    60000
74892    15000
74893     9000
74894     9200
74895    15000
Name: Rent, Length: 74896, dtype: int64

In [19]:
data.Rent.isna().sum()

0

In [20]:
data.Rent.dtypes

dtype('int64')

-   The data type of Rent  is not valid
-   In rent column data contain alphanumeric values like 1.5L,3.5L values
-   we need to convert this into the numeric value
-   This feature contain zero null value



#### Area

In [21]:
data.Area.dtypes

dtype('int64')

In [22]:
data.Area.unique()

array([1132, 1200, 1700, ..., 2604, 1548,  506])

In [23]:
data.Area.value_counts()

Area
1100    4689
1200    4379
1000    3941
550     2600
1300    2518
        ... 
1522       1
2935       1
1601       1
1095       1
2780       1
Name: count, Length: 1148, dtype: int64

-   The data type of Rent  is  valid
-   Does not contains any null value
-   it content many outlier values


#### Appartment_type

In [24]:
data.appartment_type.dtypes

dtype('int64')

In [25]:
data.appartment_type.isna().sum()

0

In [26]:
data.appartment_type.value_counts()

appartment_type
2    26250
3    23020
1    19291
4     5065
5     1138
6      125
7        4
8        2
9        1
Name: count, dtype: int64

-   The data type of Rent  is  valid
-   Does not contains any null value
-   we are going to remove the apparttment type which has ocuurance  less than 150 

In [27]:
data.Deposit.unique()


array(['No Deposit', '4 L deposit', '1.25 L deposit', '1 L deposit',
       '60,000 deposit', '70,000 deposit', '3 L deposit', '3.5 L deposit',
       '1.5 L deposit', '3.3 L deposit', '2 L deposit', '6 L deposit',
       '9.75 L deposit', '10 L deposit', '9 L deposit', '80,000 deposit',
       '42,000 deposit', '75,000 deposit', '5 L deposit',
       '40,000 deposit', '2.5 L deposit', '5.4 L deposit', '20 L deposit',
       '15 L deposit', '18 L deposit', '8.5 L deposit', '2.1 L deposit',
       '1.74 L deposit', '1.35 L deposit', '28,000 deposit',
       '1.67 L deposit', '1.2 L deposit', '1.16 L deposit',
       '9.9 L deposit', '3.6 L deposit', '6.2 L deposit', '2.2 L deposit',
       '4.8 L deposit', '1.6 L deposit', '2.6 L deposit', '1.7 L deposit',
       '1.3 L deposit', '81,000 deposit', '84,000 deposit',
       '69,000 deposit', '1.08 L deposit', '1.02 L deposit',
       '1.56 L deposit', '54,000 deposit', '1.05 L deposit',
       '50,000 deposit', '7,000 deposit', '30,000 de

In [28]:
(
    data
    .Deposit
    .loc[lambda ser: ser.str.contains("[^No Deposit]")]
     .count()
    
)

9393

-   The data type of Deposit   is not  valid
-   This feature have lots of value as "no_Deposit"
-   because of that we are going to drop this feature

#### Bathrooms	

In [29]:
data.Bathrooms	

0        2 bathrooms
1        2 bathrooms
2        3 bathrooms
3        2 bathrooms
4        3 bathrooms
            ...     
74891    3 bathrooms
74892    2 bathrooms
74893    1 bathrooms
74894    1 bathrooms
74895    1 bathrooms
Name: Bathrooms, Length: 74896, dtype: object

In [30]:
data.Bathrooms.isna().sum()

344

In [31]:
(
    data
    .Bathrooms
     .str.strip()
     .str.get(0)
      
     
     
    
)

0        2
1        2
2        3
3        2
4        3
        ..
74891    3
74892    2
74893    1
74894    1
74895    1
Name: Bathrooms, Length: 74896, dtype: object

-   The data type of Bathrooms is not  valid
-   it should be numeric
-   this feature contains 344 null value
-   in datacleaning operation the above problems are handle

#### Facing


-   This feature have lots of null value
-   because of that we are going to drop this feature

#### Address

In [32]:
data.Address

0                  Ulwe, Mumbai
1                  Ulwe, Mumbai
2                  Ulwe, Mumbai
3        Kandivali West, Mumbai
4               Chembur, Mumbai
                  ...          
74891          Sector 79, Noida
74892         Sector 128, Noida
74893          Sector 52, Noida
74894         Sector 135, Noida
74895          Sector 44, Noida
Name: Address, Length: 74896, dtype: object

In [33]:
data.Address.isna().sum()

0

In [34]:
data['Address'].str.split(',').apply(lambda x:x[1]).unique()

array([' Mumbai', '', 'Thane west', ' Pune', ' Bangalore', ' Chennai',
       ' Delhi', ' Gurgaon', ' Hyderabad', ' Kolkata', ' Ahmedabad',
       ' Noida'], dtype=object)

In [35]:
len(data['Address'].str.split(',').apply(lambda x:x[1]).unique())

12

In [36]:
len(data['Address'].str.split(',').apply(lambda x:x[0]).unique())

498


-   This  Address feature have loction and city 
-   we have requirement of city name
-   because of that we have to seprate out the locality and city from address
-   add new feature as city and drop the feature address from dataframe
-   This feature does not contains missing values 

#### Apartment Name 

-   This feature have lots of null value
-   because of that we are going to drop this feature

#### Status

In [37]:
data.Status

0           Unfurnished
1           Unfurnished
2        Semi-Furnished
3           Unfurnished
4        Semi-Furnished
              ...      
74891    Semi-Furnished
74892    Semi-Furnished
74893         Furnished
74894         Furnished
74895         Furnished
Name: Status, Length: 74896, dtype: object

In [38]:
data.Status.isna().sum()

0

In [39]:
data.Status.unique()

array(['Unfurnished', 'Semi-Furnished', 'Furnished'], dtype=object)

-   This Feature has valid datatype
-   Does not contains any null value
-   not required any cleaning opertion

#### Rent/sqfit

-   There is no use of this feature 
-   because of that we are going to drop this feature

### Locality_class

In [40]:
data.localityclass.isna().sum()


0

-   This Feature has valid datatype
-   Does not contains any null value
-   not required any cleaning opertion

### 5.DATA CLEANING OPERATION


In [41]:
 def convert_rent(rent):
    if pd.isna(rent):  # Check for NaN values
        return np.nan
    
    rent = str(rent)  # Ensure rent is treated as a string
    
    if "L" in rent:
        return int(float(rent.replace('L', '').replace(',', '').strip()) * 100000)
    else:
        return int(float(rent.replace(',', '').strip()))

In [42]:
def clean_data(df):
    # Drop the 'broker_consultancy' column and rename columns to lowercase
    df_cleaned = (
        df
         
        #.drop(columns="broker_consultancy")  # Drop the 'broker_consultancy','Deposit' column
        .drop(columns="Deposit")
        .drop(columns="Facing")
        #.drop(columns="Apartment Name")
        #.drop(columns="Rent/sqfit")
        
        #.drop_duplicates()
        
        .rename(columns=str.lower)            # Rename columns to lowercase
        .assign
        ( bathrooms=lambda df_:
         (
            df_
            
         .bathrooms
         .str.strip()
          .str.get(0)   
             
         )
            
        )
   
    )

    

    

    # Apply the convert_rent function to the 'rent' column
    df_cleaned['rent'] = df_cleaned['rent'].apply(convert_rent)

    if 'bathrooms' in df_cleaned.columns:
        mode_bathrooms = df_cleaned['bathrooms'].mode()[0]  # Calculate mode
        df_cleaned['bathrooms'] = df_cleaned['bathrooms'].fillna(mode_bathrooms) 

    # Convert 'bathrooms' column to integer type
    df_cleaned['bathrooms'] = df_cleaned['bathrooms'].astype(int)

    df_cleaned['city']=df_cleaned['address'].str.split(',').apply(lambda x:x[1])
    df_cleaned['locality']=df_cleaned['address'].str.split(',').apply(lambda x:x[0])
    
    df_cleaned = df_cleaned.drop(columns="address")
    
    # Rename columns by removing or replacing substrings
    #df_cleaned.columns = df_cleaned.columns.str.replace('area', '')
    #df_cleaned.columns = df_cleaned.columns.str.replace('builtup', 'size')
    df_cleaned.columns = df_cleaned.columns.str.replace('appartment_type', 'type')
    df_cleaned.columns = df_cleaned.columns.str.replace('status', 'furnished_type')

    if 'type' in df_cleaned.columns:
        value_counts = df_cleaned['type'].value_counts()
        types_to_keep = value_counts[value_counts >= 150].index
        df_cleaned = df_cleaned[df_cleaned['type'].isin(types_to_keep)]

    df_cleaned['city'] =  df_cleaned['city'].str.strip()
    df_cleaned['city'] = df_cleaned['city'].replace('Thane west', 'Mumbai')
    df_cleaned['city'] = df_cleaned['city'].replace('','Mumbai')
    
    

    
     
    

    


    


    
    




    

    return df_cleaned
    


In [43]:
cleandata=clean_data(data)


In [44]:
cleandata.info()

<class 'pandas.core.frame.DataFrame'>
Index: 74764 entries, 0 to 74895
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   broker_consultancy  74763 non-null  object 
 1   rent                74764 non-null  int64  
 2   area                74764 non-null  int64  
 3   type                74764 non-null  int64  
 4   bathrooms           74764 non-null  int64  
 5   apartment name      30344 non-null  object 
 6   furnished_type      74764 non-null  object 
 7   rent/sqfit          74764 non-null  float64
 8   localityclass       74764 non-null  object 
 9   city                74764 non-null  object 
 10  locality            74764 non-null  object 
dtypes: float64(1), int64(4), object(6)
memory usage: 6.8+ MB


In [45]:
cleandata.describe()

rent     area     type  bathrooms  rent/sqfit
count   74764.00 74764.00 74764.00   74764.00    74764.00
mean    40850.29  1241.72     2.23       2.16       32.61
std     51619.00   753.79     0.96       0.98       25.36
min      1600.00   150.00     1.00       1.00        4.00
25%     15000.00   750.00     1.00       1.00       16.67
50%     26000.00  1100.00     2.00       2.00       25.15
75%     45000.00  1500.00     3.00       3.00       38.18
max   1800000.00  8326.00     5.00       7.00      470.00

In [46]:
cleandata.sample(150)

broker_consultancy   rent  area  type  bathrooms  \
5805   GN Properties\nAGENT\n-  25000  1200     2          2   
46889  Gaurav Dahiya\nAGENT\n-  28000  1100     3          2   
24598   Seller\nVERIFIED OWNER  20000   550     1          1   
71149   Seller\nVERIFIED OWNER  15000  1125     2          2   
64783   Seller\nVERIFIED OWNER   9500   150     1          1   
...                        ...    ...   ...   ...        ...   
55407  Investors Hub\nAGENT\n-  95000  3070     4          4   
14699            Seller\nOWNER  25000  1200     2          2   
17373    KDH REALITY\nAGENT\n0  24000  1250     2          2   
29735   Seller\nVERIFIED OWNER   6500   350     1          1   
52091   Seller\nVERIFIED OWNER  22000   700     1          1   

                             apartment name  furnished_type  rent/sqfit  \
5805          Reputed Builder Madhav Shruti       Furnished       20.83   
46889                                   NaN  Semi-Furnished       25.45   
24598                                   NaN       Furnished       36.36   
71149                                   NaN  Semi-Furnished       13.33   
64783                                   NaN       Furnished       63.33   
...                                     ...             ...         ...   
55407                     Aparna Luxor Park  Semi-Furnished       30.94   
14699                      Godrej The Trees  Semi-Furnished       20.83   
17373  Kolte Patil Ivy Apartment E V And VI  Semi-Furnished       19.20   
29735                                   NaN  Semi-Furnished       18.57   
52091                                   NaN       Furnished       31.43   

      localityclass       city        locality  
5805         class2     Mumbai     Kalyan West  
46889        class2      Delhi  Mansa Ram Park  
24598        class1  Bangalore      Whitefield  
71149        class3  Ahmedabad      Hatkeshwar  
64783        class1    Kolkata       Sector 22  
...             ...        ...             ...  
55407        class1  Hyderabad        Kondapur  
14699        class2     Mumbai        Vikhroli  
17373        class2       Pune         Wagholi  
29735        class2  Bangalore     J. P. Nagar  
52091        class1    Gurgaon   Sushant LOK I  

[150 rows x 11 columns]

In [47]:
len(cleandata.city.unique())

10

In [48]:
cleandata.city.value_counts()

city
Mumbai       16085
Delhi        14900
Hyderabad     8680
Bangalore     8246
Pune          7220
Gurgaon       6060
Ahmedabad     4760
Kolkata       3960
Noida         2820
Chennai       2033
Name: count, dtype: int64

In [49]:
cleandata.type.value_counts()

type
2    26250
3    23020
1    19291
4     5065
5     1138
Name: count, dtype: int64

In [50]:
cleandata.localityclass.value_counts()

localityclass
class2    33137
class1    24162
class3    17465
Name: count, dtype: int64

### 6.OUTLIERS DETECTIONS

In [51]:

def find_outliers(df, column_name):
    """
    Identifies outliers in the specified column of the DataFrame using the IQR method.
    """
    # Check if the column exists in the DataFrame
    if column_name not in df.columns:
        raise KeyError(f"Column '{column_name}' does not exist in the DataFrame.")
    
    # Calculate Q1 (25th percentile) and Q3 (75th percentile)
    Q1 = df[column_name].quantile(0.25)
    Q3 = df[column_name].quantile(0.75)
    # Calculate IQR
    IQR = Q3 - Q1
    # Determine outlier bounds
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    # Return boolean mask for outliers
    return (df[column_name] < lower_bound) | (df[column_name] > upper_bound)


#### Rent

In [52]:

# Find outliers for 'rent'
rent_outliers = find_outliers(cleandata, 'rent')
print("Outliers for 'rent':")
print(rent_outliers)

Outliers for 'rent':
0        False
1        False
2        False
3        False
4        False
         ...  
74891    False
74892    False
74893    False
74894    False
74895    False
Name: rent, Length: 74764, dtype: bool


In [53]:
def drop_outliers(df, column_name):
    """
    Drops outliers from the specified column of the DataFrame and returns the cleaned DataFrame.
    """
    # Find outliers
    outlier_mask = find_outliers(df, column_name)
    
    # Ensure that outlier_mask is a boolean Series
    if not pd.api.types.is_bool_dtype(outlier_mask):
        raise TypeError("The outlier mask must be a boolean Series.")
    
    # Drop rows where the outlier mask is True
    df_cleaned = df[~outlier_mask]
    return df_cleaned

In [54]:
finaldata=drop_outliers(cleandata,'rent')

In [55]:
merge_data=drop_outliers(cleandata,'rent')


In [56]:
# Save the DataFrame to a CSV file
merge_data.to_csv('visul.csv', index=False)


In [48]:
finaldata

rent  area  type  bathrooms  furnished_type localityclass    city
0      20000  1132     2          2     Unfurnished        class2  Mumbai
1      20000  1200     2          2     Unfurnished        class2  Mumbai
2      30000  1700     3          3  Semi-Furnished        class2  Mumbai
3      28500   750     2          2     Unfurnished        class2  Mumbai
4      50000  1400     3          3  Semi-Furnished        class2  Mumbai
...      ...   ...   ...        ...             ...           ...     ...
74891  60000  1880     3          3  Semi-Furnished        class1   Noida
74892  15000   500     2          2  Semi-Furnished        class2   Noida
74893   9000   150     1          1       Furnished        class1   Noida
74894   9200   300     1          1       Furnished        class1   Noida
74895  15000   225     1          1       Furnished        class1   Noida

[69567 rows x 7 columns]

In [49]:
finaldata.describe()

rent     area     type  bathrooms
count 69567.00 69567.00 69567.00   69567.00
mean  30354.45  1142.10     2.15       2.06
std   20129.71   618.07     0.92       0.91
min    1600.00   150.00     1.00       1.00
25%   15000.00   700.00     1.00       1.00
50%   25000.00  1020.00     2.00       2.00
75%   41000.00  1340.00     3.00       2.00
max   90000.00  4200.00     5.00       6.00

In [59]:
# Save the DataFrame to a CSV file
finaldata.to_csv('output.csv', index=False)


### 7.Preprocessing Operations

In [50]:
finaldata

rent  area  type  bathrooms  furnished_type localityclass    city
0      20000  1132     2          2     Unfurnished        class2  Mumbai
1      20000  1200     2          2     Unfurnished        class2  Mumbai
2      30000  1700     3          3  Semi-Furnished        class2  Mumbai
3      28500   750     2          2     Unfurnished        class2  Mumbai
4      50000  1400     3          3  Semi-Furnished        class2  Mumbai
...      ...   ...   ...        ...             ...           ...     ...
74891  60000  1880     3          3  Semi-Furnished        class1   Noida
74892  15000   500     2          2  Semi-Furnished        class2   Noida
74893   9000   150     1          1       Furnished        class1   Noida
74894   9200   300     1          1       Furnished        class1   Noida
74895  15000   225     1          1       Furnished        class1   Noida

[69567 rows x 7 columns]

-   This Dataset have city and furnished_type,localityclass  as categorical variable
-   we have to transform into the numeric value using Label_Encoder

In [51]:
from sklearn.preprocessing import LabelEncoder
encoder_city=LabelEncoder()
encoder_class=LabelEncoder()
encoder_furnished=LabelEncoder()


In [56]:
finaldata['city']= encoder_city.fit_transform(finaldata['city'])

# Retrieve the mapping of cities to encoded values
city_mapping = dict(zip(encoder_city.classes_, encoder_city.transform(encoder_city.classes_)))

# Print the mapping
print("\nCity to Encoded Value Mapping:")
print(city_mapping)


City to Encoded Value Mapping:
{'Ahmedabad': 0, 'Bangalore': 1, 'Chennai': 2, 'Delhi': 3, 'Gurgaon': 4, 'Hyderabad': 5, 'Kolkata': 6, 'Mumbai': 7, 'Noida': 8, 'Pune': 9}


/tmp/ipykernel_11811/2800574516.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  finaldata['city']= encoder_city.fit_transform(finaldata['city'])


In [57]:
finaldata['furnished_type']=encoder_furnished.fit_transform(finaldata['furnished_type'])

# Retrieve the mapping of cities to encoded values
furnished_mapping = dict(zip(encoder_furnished.classes_, encoder_furnished.transform(encoder_furnished.classes_)))

# Print the mapping
print("\nfurnished  to Encoded Value Mapping:")
print(furnished_mapping)


furnished  to Encoded Value Mapping:
{'Furnished': 0, 'Semi-Furnished': 1, 'Unfurnished': 2}


/tmp/ipykernel_11811/2027320929.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  finaldata['furnished_type']=encoder_furnished.fit_transform(finaldata['furnished_type'])


In [58]:
finaldata['localityclass']=encoder_class.fit_transform(finaldata['localityclass'])

# Retrieve the mapping of cities to encoded values
class_mapping = dict(zip(encoder_class.classes_, encoder_class.transform(encoder_class.classes_)))

# Print the mapping
print("\nlocality to Encoded Value Mapping:")
print(class_mapping)


locality to Encoded Value Mapping:
{'class1': 0, 'class2': 1, 'class3': 2}


/tmp/ipykernel_11811/844390390.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  finaldata['localityclass']=encoder_class.fit_transform(finaldata['localityclass'])


In [59]:
finaldata

rent  area  type  bathrooms  furnished_type  localityclass  city
0      20000  1132     2          2               2              1     7
1      20000  1200     2          2               2              1     7
2      30000  1700     3          3               1              1     7
3      28500   750     2          2               2              1     7
4      50000  1400     3          3               1              1     7
...      ...   ...   ...        ...             ...            ...   ...
74891  60000  1880     3          3               1              0     8
74892  15000   500     2          2               1              1     8
74893   9000   150     1          1               0              0     8
74894   9200   300     1          1               0              0     8
74895  15000   225     1          1               0              0     8

[69567 rows x 7 columns]

In [60]:
finaldata

rent  area  type  bathrooms  furnished_type  localityclass  city
0      20000  1132     2          2               2              1     7
1      20000  1200     2          2               2              1     7
2      30000  1700     3          3               1              1     7
3      28500   750     2          2               2              1     7
4      50000  1400     3          3               1              1     7
...      ...   ...   ...        ...             ...            ...   ...
74891  60000  1880     3          3               1              0     8
74892  15000   500     2          2               1              1     8
74893   9000   150     1          1               0              0     8
74894   9200   300     1          1               0              0     8
74895  15000   225     1          1               0              0     8

[69567 rows x 7 columns]

In [62]:
finaldata

rent  area  type  bathrooms  furnished_type  localityclass  city
0      20000  1132     2          2               2              1     7
1      20000  1200     2          2               2              1     7
2      30000  1700     3          3               1              1     7
3      28500   750     2          2               2              1     7
4      50000  1400     3          3               1              1     7
...      ...   ...   ...        ...             ...            ...   ...
74891  60000  1880     3          3               1              0     8
74892  15000   500     2          2               1              1     8
74893   9000   150     1          1               0              0     8
74894   9200   300     1          1               0              0     8
74895  15000   225     1          1               0              0     8

[69567 rows x 7 columns]

In [63]:
finaldata.corr()

rent  area  type  bathrooms  furnished_type  localityclass  \
rent            1.00  0.58  0.50       0.58           -0.09          -0.43   
area            0.58  1.00  0.75       0.82           -0.02           0.17   
type            0.50  0.75  1.00       0.82            0.01           0.17   
bathrooms       0.58  0.82  0.82       1.00            0.09           0.07   
furnished_type -0.09 -0.02  0.01       0.09            1.00           0.27   
localityclass  -0.43  0.17  0.17       0.07            0.27           1.00   
city            0.18 -0.01  0.01       0.03            0.19          -0.21   

                city  
rent            0.18  
area           -0.01  
type            0.01  
bathrooms       0.03  
furnished_type  0.19  
localityclass  -0.21  
city            1.00

In [64]:
model_data= finaldata.sample(frac=1, random_state=42).reset_index(drop=True)


In [65]:
model_data

rent  area  type  bathrooms  furnished_type  localityclass  city
0      16000  1100     2          2               1              2     5
1      24500   550     1          1               0              0     5
2      20000  1800     3          3               0              2     5
3      11000  1250     2          1               1              2     1
4      38000  1150     2          2               1              1     7
...      ...   ...   ...        ...             ...            ...   ...
69562  35000  1200     3          2               0              1     3
69563  45000   600     2          1               1              0     7
69564  56000  1885     3          3               1              0     5
69565  70000   923     2          2               1              0     7
69566  41000  1250     2          2               1              0     9

[69567 rows x 7 columns]

### 8.Split the data

In [66]:
X = model_data.drop(columns="rent")
y = model_data.rent.copy()

In [67]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [68]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(55653, 6) (55653,)
(13914, 6) (13914,)


### 9.Train The model & model selection


In [69]:
from sklearn.linear_model import LinearRegression

model_lr = LinearRegression()
model_lr.fit(X_train, y_train)

LinearRegression()

In [70]:
from sklearn.metrics import r2_score
r2 = r2_score(y_test, model_lr.predict(X_test))
r2

0.6550482931311402

In [71]:
# get the predicted data
y_pred = model_lr.predict(X_test)

# getting the true values
y_true = y_test

In [72]:
from sklearn.metrics import mean_absolute_error

# mean absolute error (MAE)
mae = mean_absolute_error(y_true, y_pred)
mae

8910.910154525462

In [73]:
from sklearn.metrics import mean_squared_error

# mean squared error (MSE)
mse = mean_squared_error(y_true, y_pred)
mse

143540342.47448048

In [74]:
from sklearn.metrics import r2_score

# R-squared score
r2 = r2_score(y_true, y_pred)
r2

0.6550482931311402

#### polynomial Regression 

In [75]:
from sklearn.preprocessing import PolynomialFeatures

# create a encoder object
polynomial_features = PolynomialFeatures(degree=5)

# preprocess the data to get the polynomials of X
x_polynomial = polynomial_features.fit_transform(X_train)
x_polynomial

array([[1.0000e+00, 1.5000e+03, 3.0000e+00, ..., 7.2900e+02, 6.5610e+03,
        5.9049e+04],
       [1.0000e+00, 1.8860e+03, 3.0000e+00, ..., 1.2500e+02, 6.2500e+02,
        3.1250e+03],
       [1.0000e+00, 6.5000e+02, 1.0000e+00, ..., 0.0000e+00, 0.0000e+00,
        3.1250e+03],
       ...,
       [1.0000e+00, 1.0820e+03, 2.0000e+00, ..., 5.1200e+02, 4.0960e+03,
        3.2768e+04],
       [1.0000e+00, 1.1000e+03, 3.0000e+00, ..., 2.7000e+01, 8.1000e+01,
        2.4300e+02],
       [1.0000e+00, 8.5000e+02, 1.0000e+00, ..., 0.0000e+00, 0.0000e+00,
        1.0240e+03]])

In [76]:
### create Linear Regression Model
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(x_polynomial, y_train)

LinearRegression()

In [78]:
# preprocess the data to get the polynomials of X
x_polynomial_test = polynomial_features.fit_transform(X_test)
x_polynomial_test

array([[1.0000e+00, 5.0000e+02, 1.0000e+00, ..., 7.2900e+02, 6.5610e+03,
        5.9049e+04],
       [1.0000e+00, 2.2500e+02, 1.0000e+00, ..., 0.0000e+00, 0.0000e+00,
        3.2768e+04],
       [1.0000e+00, 1.2500e+03, 2.0000e+00, ..., 7.2900e+02, 6.5610e+03,
        5.9049e+04],
       ...,
       [1.0000e+00, 1.6300e+03, 3.0000e+00, ..., 0.0000e+00, 0.0000e+00,
        5.9049e+04],
       [1.0000e+00, 1.5000e+02, 1.0000e+00, ..., 0.0000e+00, 0.0000e+00,
        0.0000e+00],
       [1.0000e+00, 1.3000e+03, 3.0000e+00, ..., 7.2900e+02, 6.5610e+03,
        5.9049e+04]])

In [84]:
from sklearn.metrics import r2_score

r2 = r2_score(y_test, model.predict(x_polynomial_test))
r2

0.3718851997959127

In [79]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import AdaBoostRegressor, RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from xgboost import XGBRegressor
from catboost import CatBoostRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.svm import SVR
from sklearn.naive_bayes import GaussianNB
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline


In [80]:
# Initialize models
models = {
    'Linear Regression': LinearRegression(),
    'Decision Tree': DecisionTreeRegressor(random_state=42),
    'XGBoost': XGBRegressor(random_state=42),
    'CatBoost': CatBoostRegressor(silent=True, random_state=42),
    'AdaBoost': AdaBoostRegressor(random_state=42),
    'Random Forest': RandomForestRegressor(random_state=42),
    
}

# Train and evaluate models
results = {}

for name, model in models.items():
    # Train the model
    model.fit(X_train, y_train)
    
    # Make predictions
    y_pred = model.predict(X_test)
    
    # Evaluate the model
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    mae = mean_absolute_error(y_test, y_pred)
    
    # Store the results
    results[name] = {
        'Mean Squared Error': mse,
        'R-squared': r2,
        'Mean Absolute Error': mae
    }

# Print the evaluation results
print("Model Evaluation Report:")
for name, metrics in results.items():
    print(f"\n{name}:")
    for metric, value in metrics.items():
        print(f"{metric}: {value:.4f}")

Model Evaluation Report:

Linear Regression:
Mean Squared Error: 143540342.4745
R-squared: 0.6550
Mean Absolute Error: 8910.9102

Decision Tree:
Mean Squared Error: 23157801.8469
R-squared: 0.9443
Mean Absolute Error: 1680.8996

XGBoost:
Mean Squared Error: 19767891.5043
R-squared: 0.9525
Mean Absolute Error: 1738.6791

CatBoost:
Mean Squared Error: 19621711.9702
R-squared: 0.9528
Mean Absolute Error: 1846.4731

AdaBoost:
Mean Squared Error: 113488381.9625
R-squared: 0.7273
Mean Absolute Error: 8767.5762

Random Forest:
Mean Squared Error: 20384814.9515
R-squared: 0.9510
Mean Absolute Error: 1644.4866


- CatBoost would be the best choice due to its lowest MSE and highest R-squared
- CatBoost stands out as the top performer

#### Final model for prediction

In [81]:
from catboost import CatBoostRegressor, Pool
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [82]:
model = CatBoostRegressor(
    iterations=500,          # Number of boosting iterations
    depth=6,                 # Depth of each tree
    learning_rate=0.1,       # Learning rate
    l2_leaf_reg=3,           # L2 regularization coefficient
    loss_function='RMSE',    # Loss function
    cat_features=[]          # Specify categorical features if any
)

# Fit the model
model.fit(X_train, y_train, eval_set=(X_test, y_test), verbose=100)

# Make predictions
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
print(f"Mean Squared Error: {mse}")
print(f"R2_Score: {r2}")
print(f"Mean Absolute Error: {mae}")
# Feature importances
feature_importances = model.get_feature_importance()
importance_df = pd.DataFrame({
    'Feature': X_train.columns,
    'Importance': feature_importances
}).sort_values(by='Importance', ascending=False)

print("\nFeature Importances:")
print(importance_df)

0:	learn: 18618.3679534	test: 18938.2938768	best: 18938.2938768 (0)	total: 21.3ms	remaining: 10.6s
100:	learn: 4796.0128572	test: 4844.2185430	best: 4844.2185430 (100)	total: 1.3s	remaining: 5.12s
200:	learn: 4515.1096842	test: 4592.1003608	best: 4592.1003608 (200)	total: 2.45s	remaining: 3.65s
300:	learn: 4404.8733462	test: 4511.3964179	best: 4511.3964179 (300)	total: 3.69s	remaining: 2.44s
400:	learn: 4338.6857377	test: 4476.2138050	best: 4476.2138050 (400)	total: 4.92s	remaining: 1.21s
499:	learn: 4289.9234407	test: 4456.1502884	best: 4456.1502884 (499)	total: 5.97s	remaining: 0us

bestTest = 4456.150288
bestIteration = 499

Mean Squared Error: 19857275.392390177
R2_Score: 0.9522796105799469
Mean Absolute Error: 1926.2923190799913

Feature Importances:
          Feature  Importance
0            area       30.87
4   localityclass       26.23
5            city       25.89
1            type        7.42
2       bathrooms        5.72
3  furnished_type        3.87


In [85]:
import pandas as pd
import numpy as np

def evaluate_model(model, area, type, bathrooms, furnished_type, localityclass, city_encoded):
    # Create a DataFrame from the input values
    input_data = pd.DataFrame({
        'area': [area],
        'type': [type],
        'bathrooms': [bathrooms],
        'furnished_type': [furnished_type],
        'localityclass': [localityclass],
        'city': [city_encoded]
    })
    
    # Preprocessing steps (these should match the steps used during training)
    # For example:
    # - Encode categorical variables
    # - Scale numerical features
    # - Handle missing values
    
    # Here, I'm assuming some common preprocessing steps; adapt them as needed:
    # Example of one-hot encoding (if necessary):
    # input_data = pd.get_dummies(input_data, columns=['type', 'furnished_type'])
    
    # Example of scaling (if necessary):
    # from sklearn.preprocessing import StandardScaler
    # scaler = StandardScaler()
    # input_data[['area', 'bathrooms']] = scaler.fit_transform(input_data[['area', 'bathrooms']])
    
    # If the model requires specific preprocessing, make sure to apply it here.
    
    # Predict with the model
    prediction = model.predict(input_data)
    
    return prediction

# Example usage:
# Assuming `trained_model` is your trained model (e.g., a scikit-learn model)
# and that you have performed necessary preprocessing steps

# trained_model = ...  # Load or define your trained model here

# New input values
# area = 1200
# type = 'Apartment'
# bathrooms = 2
# furnished_type = 'Furnished'
# localityclass = 3
# city_encoded = 1

# prediction = evaluate_model(trained_model, area, type, bathrooms, furnished_type, localityclass, city_encoded)
# print(prediction)


In [86]:
evaluate_model(model,1200,1,1,1,1,7)

array([31692.95226454])

In [87]:
# Save the model to a file
with open('model_cat.pkl', 'wb') as file:
    pickle.dump(model, file)


In [90]:
# Save the model to a file
with open('model_lr.pkl', 'wb') as file:
    pickle.dump(model, file)